bleep-core/src/scala/bleep/GenBloopFiles.scala

```diff
@@ -328,6 +328,27 @@ object GenBloopFiles {
         case _                                            => bloop.config.Tag.Library
       }
 
+    val sourceGenerators = {
+      val sg = explodedProject.`source-generator-scripts`.values.map { sc =>
+        if (!build.scripts.contains(sc.scriptName)) {
+          sys.error(s"Unknown script ${sc.scriptName} referenced by ${crossName.value}")
+        }
+        val codegenPath = buildPaths.generatedSourcesDir(crossName)
+        val scriptCall = List("bleep", sc.scriptName.value)
+        val `this` = "$" + "this"
+        // Dereferencing all '$this' into the cross name, so that users can
+        // inject the module's name into the script.
+        val args = sc.arguments.map {
+          case `this` => crossName.value
+          case other  => other
+        }
+        val fullCall = scriptCall ++ args
+        // TODO what are the globs for ?
```

Comment:
The source globs are likely used for cache invalidation - any sources changed and the resource generator is run again.

I guess that if we pull all the transitive source directories of the script project that should work, something like this:
```scala
    val scriptSources: List[Path] =
      started.build
        .scripts(scriptName)
        .values
        .flatMap {
          case model.ScriptDef.Main(scriptProjectName, _) =>
            val withTransitive = List(scriptProjectName) ++ started.build.transitiveDependenciesFor(scriptProjectName).keys
            withTransitive.flatMap { projectName =>
              val bloopProject = started.bloopProjects(projectName)
              bloopProject.sources ++ bloopProject.resources.getOrElse(Nil)
            }
          case model.ScriptDef.Shell(_, _) =>
            throw new BleepException.Text(s"cannot use shell script as resource generator, because bleep doesn't know how to invalidate it")
        }
        .distinct

```

Reply:
As you see I think for shell scripts the user will need to provide this invalidation information themselves

Reply:
Yeah the user might want to provide invalidation information for scripts too, if scripts depend on reading files from some arbitrary folder from the file system

Reply:
Yeah, we should definitely add that as an optional element in `ScriptDef.Shell`. So if you want to use that shell script as a resource generator you need to provide a list of input files. probably a list of paths relative to the build directory - that is the root of the project

Reply:
Sorry, my point was that `ScriptDef.Main` should hold that information too, because inputs to code-generators might not be in the bloop project's sources, nor it's resources.

Reply:
That's a constraint we can decide though, and i think we should

Reply:
> That's a constraint we can decide though, and i think we should 

Too constraining for my taste, tbh. Like, say my build has a source-dependency (by means of git submodules) to pull shared specifications from another repo. I'd rather not clone that entire repo in the resources of my project, so I'd clone it at the root of my project for instance, and I'd configure my code-generator to understand where the clone is located. But then, since bloop is responsible for deciding when to call my code-generator, I'd need to tell bloop to monitor the clone for changes (for when I run `git submodule update -remote`)

Reply:
Or another example closer to my heart : in the [smithy4s](https://github.com/disneystreaming/smithy4s) project I maintain, we have a [sample specs](https://github.com/disneystreaming/smithy4s/tree/series/0.17/sampleSpecs) folder that contains smithy specifications that are used both in code-generation for test-purposes, and in documentation (which is an mdoc project). Forcing a pattern where codegen inputs would HAVE to be resources of a script project would mean that it'd be harder to share the specs between my tests and my documentation. 

I also like the fact that this folder is at the root, because of how it makes discoverability of examples easier when checking-out the project

Reply:
You can add additional resource folders! :)
```yaml
projects:
  a:
    resources: ../sampleSpecs
```

How about that?

Reply:
mmm okay yeah, I guess

Reply:
Right, so I've thought about this some more as I've resumed implementing, and I think some configuration will be required, if we want to delegate the invocation of the codegeneration script to bloop. Taking the example of proto codegenerators, I think it's pretty natural for the proto files to be located in the folder of the project that would receive the generated code. 

So what bloop configuration should be generated when that's the case ? 

* the script source-path (transitive sources/resources) doesn't suffice because that doesn't encompass the target project's folder 
* `project-folder/**` would be overkill because that'd encompass the Scala sources and you probably don't want to re-trigger code-generation whenever an unmanaged source directory gets touched, that'd be wildly inefficient
* the resources of the project would be too prescriptive, because people may not want to package the input of the code-generators they run in end artifacts  
* you can't solve that problem in the script implementation, because bloop ultimately decides whether to run stuff or not, and therefore it may skip the codegenerators in situations where you'd want for it to run  

So my conclusion is that you have to let the user have control over the globs. 

The alternative would be to not rely on bloop for this, and have bleep invoke the code-generators itself. It'd probably make for a better user experience in the end, because you could then solve the caching within the script implementation itself, in a way that'd be convenient for the user.

Reply:
> I think it's pretty natural for the proto files to be located in the folder of the project that would receive the generated code.

Agreed, this will be the case in ~all existing projects.

> The alternative would be to not rely on bloop for this, and have bleep invoke the code-generators itself. It'd probably make for a better user experience in the end, because you could then solve the caching within the script implementation itself, in a way that'd be convenient for the user.

This is what I originally had in mind, but let's keep this as a last resort? I think from a performance perspective it's best to let bloop handle it, otherwise we'll need to ask bloop to compile in multiple BSP requests.

I agree with your reasoning. If input files for codegen live inside a folder which belongs to a project instead of the script project, the user will have to provide invalidation globs for it. There is no other way in that case.

How about this then? The default invalidation globs are that of the script project (sources/resources). We want users to put them there, because there is less configuration. The user is allowed to add additional globs to the script definition if the codegen script reads files outside the script project. Is that pretty much all we need?

Reply:
> How about this then? The default invalidation globs are that of the script project (sources/resources). We want users to put them there, because there is less configuration. The user is allowed to add additional globs to the script definition if the codegen script reads files outside the script project. Is that pretty much all we need?

I think that sounds about right for this first iteration 👍